### PR TITLE
Download and Upload using IO stream

### PR DESCRIPTION
**Pull request by iLLiCiTiT**
## Description
Added option to download file to BytesIO and upload file from BytesIO.

## Additional information
Sometimes it might be handy to be able to download content from server without storing it to a file but only to memory, or upload file from memory.

### CODE REVIEW

File: ayon_api/server_api.py [hunk 1]
```diff
@@ -1674,6 +1669,53 @@ def download_file(
...
+        # Create dummy object so the function does not have to check
+        #   'progress' variable everywhere
+        if progress is None:
+            progress = TransferProgress()
+
+        progress.set_destination_url(filepath)
+
+        dst_directory = os.path.dirname(filepath)
+        if not os.path.exists(dst_directory):
+            os.makedirs(dst_directory, exist_ok=True)
```
Comment (BigRoy) [inline ayon_api/server_api.py:1705]: When using `exist_ok` there's no need to check for existence?
```suggestion
         os.makedirs(dst_directory, exist_ok=True)
```

File: ayon_api/server_api.py [hunk 2]
```diff
@@ -2154,7 +2196,7 @@ def download_addon_private_file(
         # Filename can contain "subfolders"
         dst_dirpath = os.path.dirname(dst_filepath)
         if not os.path.exists(dst_dirpath):
-            os.makedirs(dst_dirpath)
+            os.makedirs(dst_dirpath, exist_ok=True)
```
Comment (BigRoy) [inline ayon_api/server_api.py:2199]: When using `exist_ok` there's no need to check for existence?
```suggestion
         os.makedirs(dst_directory, exist_ok=True)
```

**Comment by BigRoy** [comment]
Side note is that `ayon-python-api` on `pypi` seems listed as Python 2 and Python 3 even though this change technically makes it Python 3.2+ compatible.

**Comment by iLLiCiTiT** [comment]
> Side note is that [ayon-python-api on pypi](https://pypi.org/project/ayon-python-api/) seems listed as Python 2 and Python 3 even though this change technically makes it Python 3.2+ compatible.

Good point will do PR changing it.

**Comment by kalisp** [comment]
Code looks good, but I didn't know how to test it. 
I wanted to mock downloading dependency package to BytesIO and then saving it:
image


but it failed:
image


Maybe its just my messed up test.

**Review by kalisp — APPROVED** [review]
Last commit solved my issue. Checked `upload_file_from_stream` seems that is not suffering from same issue.

**Pull request merged**